design-proposals/node-labeller-refactor.md

```diff
@@ -0,0 +1,73 @@
+
+## Goals
+
+* **Decouple the inference functionality**: Separate capability inference from node labeling to streamline the node-labeller’s functionality. Improve testing coverage for both labeling and capability parsing, focusing on eliminating file reads in unit tests and introducing tests for affected components that previously lacked test coverage.
+* **Decouple node-labeller and VM Controller**: Deprecate the `node-labeller/api` package and replace it with libvirtxml, ensuring that the VM Controller obtains its required capabilities without relying on node-labeller.
+
+## Non-Goals
+
+* This proposal does **not** aim to extend the functionality of node-labeller or VM Controller, such as adding new labels or introducing more parsing options.
+
+## Definition of Users
+
+This change is not user-facing but affects all users indirectly. The node-labeller’s original functionality remains intact, so no user intervention is required.
+
+## Repos
+
+[KubeVirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+
+The redesigned node-labeller will focus solely on labeling nodes, with all capability inference logic moved to a separate package. This new package will still be part of virt-handler but will provide capabilities as a dependency for both node-labeller and VM Controller. 
+
+The `node-capabilities.sh` script, which currently infers capabilities using the libvirt API and stores them in `/var/lib/kubevirt-node-labeller`, will be adjusted accordingly as node-labeller will no longer handle capability inference.
+
+Node-labeller will now only use specific host-related data (e.g., CPU counters), and its structures and methods will be streamlined to reflect this simpler scope. It will no longer serve as a dependency provider for the VM Controller.
+
+The redundant `node-labeller/api/capabilities` package will be replaced with the existing libvirtxml library. All components currently relying on this package, along with their tests, will be updated to use libvirtxml, completing the decoupling process.
+
+To maintain flexibility, node-labeller's construction will be refactored to use the builder pattern, simplifying its structure while allowing for future extensibility.
+
+At its end state, node-labeller should be a lean, single-file component with core logic and relevant tests.
+
+## Update/Rollback Compatibility
+
+Since this change does not alter the core functionality of any components, updates and rollbacks will be fully compatible.
+
+## Functional Testing Approach
+
+The existing functional tests in `tests/infrastructure/node-labeller.go` cover these changes, ensuring proper functionality.
+
+# Implementation Phases
+
+The node-labeller currently uses the `loadAll` function to infer several host capabilities during its construction. This will be refactored across the following phases:
+
+## Phase I - Remove `loadHostCapabilities`
```

Comment:
I can't find this function in the code base. Was it already removed as part of this effort?

Reply:
Indeed it has: https://github.com/kubevirt/kubevirt/pull/12652